File: 21주차/양궁대회.py
def calc(a,b):
    t1 = 0
    t2 = 0# 라이언
    for i in range(11):
        if a[i] > b[i]:
            t1 += (10-i)
        elif a[i] < b[i]:
            t2 += (10-i)
        elif a[i] == b[i] and a[i] == 0:
            continue
        elif a[i] == b[i] and a[i] != 0:
            t1 += (10-i)
    return t2-t1
def solution(n,info):
    ans = []
    arr = []
    def dfs(L,total):
        if L == 11:
            if total < n:
                arr[10] = n-total

            diff = calc(info,arr)
            if diff <= 0:
                ans.append([-1]+arr)
            else:
                ans.append([diff]+arr)
            
            return
        
        if total + info[L]+1 <= n:
            arr.append(info[L]+1)
            dfs(L+1,total+info[L]+1)
            arr.pop()
        arr.append(0)
        dfs(L+1,total)
        arr.pop()
    dfs(0,0)
    
    ans.sort(key=lambda x: (-x[0],-x[10],-x[9],-x[8],-x[7],-x[6],-x[5],-x[4],-x[3],-x[2],-x[1],-x[0]))
    
    if ans[0][0] == -1:
        return [-1]
    return ans[0][1:]

File: 51주차/양궁대회.js
function solution(n, info) {
  var answer = [];
  const dfs = (L, total, a, r, arr) => {
    if (L === 11) {
      // 마지막 발까지 갔다면, 라이언 점수가 더 높으면 answer에 후보 넣기.
      if (a < r) answer.push([r - a, ...arr]);
      return;
    }
    // info[L]+1 or 0 발을 쏘기 때문에 마지막에 총 n발 개수를 맞춰야 함!
    if (total + info[L] + 1 <= n) {
      // 어피치보다 +1발 더쏴도 n발 넘게 쏘지 않는다면
      dfs(L + 1, total + info[L] + 1, a, r + 10 - L, [...arr, info[L] + 1]);
    }
    // 쏘지 않기. 어피치가 이김(하지만 어피치가 0개를 쐈다면 점수 없음.)
    if (info[L] > 0) dfs(L + 1, total, a + 10 - L, r, [...arr, 0]);
    else dfs(L + 1, total, a, r, [...arr, 0]);
  };
  dfs(0, 0, 0, 0, []);
  if (answer.length == 0) return [-1]; // 이길 수 없는 경우.
  answer.sort((a, b) => {
    if (a[0] !== b[0]) return b[0] - a[0]; // 점수차이가 큰 순.
    for (let i = 11; i >= 1; i--) {
      // 차례대로 낮은점수가 많은 순.
      if (a[i] !== b[i]) return b[i] - a[i];
    }
  });

  answer = answer[0].slice(1, 12); // 각 점수 몇 발 쐈는지 배열 반환.
  let sum = answer.reduce((acc, cur) => {
    return (acc += cur);
  }, 0);
  if (sum < n) answer[answer.length - 1] = n - sum; // 만약 정답에서 총 n발보다 작으면 0점에 추가.
  return answer;
}
// 프로그래머스
